// react-native-gifted-charts/dist/BubbleChart/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, Fragment as _Fragment, jsxs as _jsxs } from "react/jsx-runtime";
import { LineDefaults, useBubbleChart, } from 'gifted-charts-core';
import BarAndLineChartsWrapper from '../Components/BarAndLineChartsWrapper';
import { Fragment, useCallback, useEffect, useMemo, useRef } from 'react';
import { Animated, Easing, I18nManager, Text, View, } from 'react-native';
import { Circle, ForeignObject, Rect, Svg, Text as CanvasText, } from 'react-native-svg';
import { isWebApp, screenWidth } from '../utils';
var AnimatedCircle = Animated.createAnimatedComponent(Circle);
var AnimatedRect = Animated.createAnimatedComponent(Rect);
export var BubbleChart = function (props) {
    var _a, _b, _c, _d;
    var opacityValue = useMemo(function () { return new Animated.Value(0); }, []);
    var pointsOpacityValue = useMemo(function () { return new Animated.Value(0); }, []);
    var secondaryXAxis = props.secondaryXAxis;
    var _e = useBubbleChart(__assign(__assign({}, props), { parentWidth: (_a = props.parentWidth) !== null && _a !== void 0 ? _a : screenWidth })), barAndLineChartsWrapperProps = _e.barAndLineChartsWrapperProps, totalWidth = _e.totalWidth, animationDuration = _e.animationDuration, _f = _e.containerHeightIncludingBelowXAxis, containerHeightIncludingBelowXAxis = _f === void 0 ? 0 : _f, getY = _e.getY, getX = _e.getX, maxValue = _e.maxValue, selectedIndex = _e.selectedIndex, setSelectedIndex = _e.setSelectedIndex, showTextOnFocus = _e.showTextOnFocus, focusEnabled = _e.focusEnabled, focusTogether = _e.focusTogether, selectedLineNumber = _e.selectedLineNumber, lastLineNumber = _e.lastLineNumber, initialSpacing = _e.initialSpacing, spacing = _e.spacing, containerHeight = _e.containerHeight, handleFocus = _e.handleFocus, handleUnFocus = _e.handleUnFocus, isAnimated = _e.isAnimated, showDataPointOnFocus = _e.showDataPointOnFocus, showDataPointLabelOnFocus = _e.showDataPointLabelOnFocus, dataPointsShape = _e.dataPointsShape, dataPointsWidth = _e.dataPointsWidth, dataPointsHeight = _e.dataPointsHeight, dataPointsColor = _e.dataPointsColor, dataPointsRadius = _e.dataPointsRadius, textColor = _e.textColor, textFontSize = _e.textFontSize, startIndex = _e.startIndex, endIndex = _e.endIndex, showValuesAsDataPointsText = _e.showValuesAsDataPointsText, cumulativeSpacing = _e.cumulativeSpacing, hideDataPoints = _e.hideDataPoints, xAxisLabelsVerticalShift = _e.xAxisLabelsVerticalShift, labelsExtraHeight = _e.labelsExtraHeight, xAxisThickness = _e.xAxisThickness, xAxisTextNumberOfLines = _e.xAxisTextNumberOfLines, rotateLabel = _e.rotateLabel, allowFontScaling = _e.allowFontScaling, borderColor = _e.borderColor, borderWidth = _e.borderWidth, opacity = _e.opacity;
    var radiiValues = useMemo(function () { var _a; return ((_a = props.data) !== null && _a !== void 0 ? _a : []).map(function (_) { return new Animated.Value(0); }); }, []);
    var pointsHeightValues = useMemo(function () { var _a; return ((_a = props.data) !== null && _a !== void 0 ? _a : []).map(function (_) { return new Animated.Value(0); }); }, []);
    var pointsWidthValues = useMemo(function () { var _a; return ((_a = props.data) !== null && _a !== void 0 ? _a : []).map(function (_) { return new Animated.Value(0); }); }, []);
    var scrollRef = (_b = props.scrollRef) !== null && _b !== void 0 ? _b : useRef(null);
    var widthValue = useMemo(function () { return new Animated.Value(0); }, []);
    var appearingOpacity = opacityValue.interpolate({
        inputRange: [0, 1],
        outputRange: [0, 1],
    });
    var growingRadii = radiiValues.map(function (r, i) {
        var _a, _b;
        return r.interpolate({
            inputRange: [0, 1],
            outputRange: [0, (_b = (_a = props.data) === null || _a === void 0 ? void 0 : _a[i].r) !== null && _b !== void 0 ? _b : dataPointsRadius],
        });
    });
    var growingHeight = pointsHeightValues.map(function (h, i) {
        var _a, _b;
        return h.interpolate({
            inputRange: [0, 1],
            outputRange: [0, (_b = (_a = props.data) === null || _a === void 0 ? void 0 : _a[i].dataPointHeight) !== null && _b !== void 0 ? _b : dataPointsHeight],
        });
    });
    var growingWidth = pointsHeightValues.map(function (w, i) {
        var _a, _b;
        return w.interpolate({
            inputRange: [0, 1],
            outputRange: [0, (_b = (_a = props.data) === null || _a === void 0 ? void 0 : _a[i].dataPointWidth) !== null && _b !== void 0 ? _b : dataPointsWidth],
        });
    });
    var appearingDataPoints = pointsOpacityValue.interpolate({
        inputRange: [0, 1],
        outputRange: [0, opacity],
    });
    var decreaseWidth = useCallback(function () {
        widthValue.setValue(0);
        Animated.timing(widthValue, {
            toValue: totalWidth,
            duration: animationDuration,
            easing: Easing.linear,
            useNativeDriver: false,
        }).start();
    }, [animationDuration, widthValue]);
    var labelsAppear = useCallback(function () {
        opacityValue.setValue(0);
        Animated.timing(opacityValue, {
            toValue: 1,
            duration: 500,
            easing: Easing.ease,
            useNativeDriver: false,
        }).start();
    }, [opacityValue]);
    var radiiGrow = useCallback(function () {
        radiiValues.forEach(function (radiusValue) {
            radiusValue.setValue(0);
            Animated.timing(radiusValue, {
                toValue: 1,
                duration: animationDuration,
                easing: Easing.ease,
                useNativeDriver: false,
            }).start();
        });
    }, [radiiValues]);
    var dataPointsHeightsGrow = useCallback(function () {
        pointsHeightValues.forEach(function (h) {
            h.setValue(0);
            Animated.timing(h, {
                toValue: 1,
                duration: animationDuration,
                easing: Easing.ease,
                useNativeDriver: false,
            }).start();
        });
    }, [pointsHeightValues]);
    var dataPointsWidthsGrow = useCallback(function () {
        pointsWidthValues.forEach(function (w) {
            w.setValue(0);
            Animated.timing(w, {
                toValue: 1,
                duration: animationDuration,
                easing: Easing.ease,
                useNativeDriver: false,
            }).start();
        });
    }, [pointsWidthValues]);
    // const dataPointsHeightGrow
    var dataPointsAppear = useCallback(function () {
        pointsOpacityValue.setValue(0);
        Animated.timing(pointsOpacityValue, {
            toValue: 1,
            duration: animationDuration,
            easing: Easing.ease,
            useNativeDriver: false,
        }).start();
    }, [pointsOpacityValue]);
    useEffect(function () {
        if (isAnimated) {
            decreaseWidth();
            labelsAppear();
            radiiGrow();
            dataPointsAppear();
            if (dataPointsShape === 'rectangular') {
                dataPointsHeightsGrow();
                dataPointsWidthsGrow();
            }
        }
    }, [isAnimated, dataPointsShape]);
    var svgHeight = containerHeightIncludingBelowXAxis + ((_c = props.overflowBottom) !== null && _c !== void 0 ? _c : 0);
    var onStripPress = function (item, index) {
        if (props.focusedDataPointIndex === undefined || !props.onFocus) {
            setSelectedIndex(index);
        }
        if (props.onFocus) {
            props.onFocus(item, index);
        }
    };
    var renderLabel = function (top, index, label, labelTextStyle, labelComponent) {
        var _a, _b;
        return (_jsx(View, { style: [
                {
                    position: 'absolute',
                    bottom: top
                        ? containerHeight +
                            60 +
                            ((_a = secondaryXAxis === null || secondaryXAxis === void 0 ? void 0 : secondaryXAxis.labelsDistanceFromXaxis) !== null && _a !== void 0 ? _a : 15)
                        : -xAxisTextNumberOfLines * 18,
                    zIndex: 10,
                    width: spacing + labelsExtraHeight,
                    left: initialSpacing + spacing * index - spacing / 2,
                    height: (_b = props.xAxisLabelsHeight) !== null && _b !== void 0 ? _b : xAxisTextNumberOfLines * 18,
                },
                rotateLabel && { transform: [{ rotate: '60deg' }] },
            ], children: labelComponent ? (labelComponent()) : (_jsx(Text, { style: [{ textAlign: 'center' }, labelTextStyle], allowFontScaling: allowFontScaling, numberOfLines: xAxisTextNumberOfLines, children: label })) }));
    };
    var renderAnimatedLabel = function (top, index, label, labelTextStyle, labelComponent) {
        var _a, _b;
        return (_jsx(Animated.View, { style: [
                {
                    height: rotateLabel
                        ? 40
                        : ((_a = props.xAxisLabelsHeight) !== null && _a !== void 0 ? _a : xAxisTextNumberOfLines * 18),
                    position: 'absolute',
                    bottom: top
                        ? containerHeight +
                            60 +
                            ((_b = secondaryXAxis === null || secondaryXAxis === void 0 ? void 0 : secondaryXAxis.labelsDistanceFromXaxis) !== null && _b !== void 0 ? _b : 15)
                        : rotateLabel
                            ? 10
                            : -xAxisTextNumberOfLines * 18,
                    zIndex: 10,
                    width: spacing,
                    left: initialSpacing + spacing * index - spacing / 2,
                    opacity: appearingOpacity,
                },
                rotateLabel && { transform: [{ rotate: '60deg' }] },
            ], children: labelComponent ? (labelComponent()) : (_jsx(Text, { allowFontScaling: allowFontScaling, style: [{ textAlign: 'center' }, labelTextStyle], numberOfLines: xAxisTextNumberOfLines, children: label })) }));
    };
    var renderDataPoints = function (hideDataPoints, dataForRender, originalDataFromProps, dataPtsShape, dataPtsWidth, dataPtsHeight, dataPtsColor, dataPtsRadius, textColor, textFontSize, startIndex, endIndex, isSecondary, showValuesAsDataPointsText, spacingArray, key) {
        var getYOrSecondaryY = getY; //isSecondary ? getSecondaryY : getY;
        return dataForRender.map(function (item, index) {
            var _a, _b, _c, _d, _e, _f, _g, _h;
            if (index < startIndex || index > endIndex)
                return null;
            if (item.hideDataPoint) {
                return null;
            }
            var dataPointsShape, dataPointsWidth, dataPointsHeight, dataPointsColor, dataPointsRadius, text, customDataPoint, dataPointLabelComponent;
            if (index === selectedIndex &&
                (focusTogether || key === selectedLineNumber)) {
                dataPointsShape =
                    item.focusedDataPointShape ||
                        props.focusedDataPointShape ||
                        item.dataPointShape ||
                        dataPtsShape;
                dataPointsWidth =
                    item.focusedDataPointWidth ||
                        props.focusedDataPointWidth ||
                        item.dataPointWidth ||
                        dataPtsWidth;
                dataPointsHeight =
                    item.focusedDataPointHeight ||
                        props.focusedDataPointHeight ||
                        item.dataPointHeight ||
                        dataPtsHeight;
                dataPointsColor =
                    item.focusedDataPointColor ||
                        props.focusedDataPointColor ||
                        LineDefaults.focusedDataPointColor;
                dataPointsRadius =
                    (_c = (_b = (_a = item.focusedDataPointRadius) !== null && _a !== void 0 ? _a : props.focusedDataPointRadius) !== null && _b !== void 0 ? _b : item.r) !== null && _c !== void 0 ? _c : dataPtsRadius;
                if (showTextOnFocus) {
                    text = item.dataPointText;
                }
                customDataPoint =
                    item.focusedCustomDataPoint ||
                        props.focusedCustomDataPoint ||
                        item.customDataPoint ||
                        props.customDataPoint;
                dataPointLabelComponent =
                    item.focusedDataPointLabelComponent ||
                        item.dataPointLabelComponent ||
                        props.focusedDataPointLabelComponent ||
                        props.dataPointLabelComponent;
            }
            else {
                dataPointsShape = item.dataPointShape || dataPtsShape;
                dataPointsWidth = item.dataPointWidth || dataPtsWidth;
                dataPointsHeight = item.dataPointHeight || dataPtsHeight;
                dataPointsColor = item.dataPointColor || dataPtsColor;
                dataPointsRadius = (_d = item.r) !== null && _d !== void 0 ? _d : dataPtsRadius;
                if (showTextOnFocus) {
                    text = '';
                }
                customDataPoint = item.customDataPoint || props.customDataPoint;
                dataPointLabelComponent =
                    item.dataPointLabelComponent || props.dataPointLabelComponent;
            }
            if (showValuesAsDataPointsText) {
                text = originalDataFromProps[index].y;
            }
            var dataPointLabelWidth = item.dataPointLabelWidth
                ? item.dataPointLabelWidth
                : props.dataPointLabelWidth
                    ? props.dataPointLabelWidth
                    : 30;
            return (_jsxs(Fragment, { children: [focusEnabled ? (_jsx(_Fragment, { children: key === lastLineNumber - 1 ? (_jsx(Rect, { x: initialSpacing + (spacing * index - spacing / 2), y: 8, width: spacing, height: containerHeight - 0, fill: 'none', onPressIn: function (evt) {
                                var locationY = evt.nativeEvent.locationY; // Note that we have another property named pageY which can be useful
                                handleFocus(index);
                            }, onPressOut: handleUnFocus })) : null })) : null, hideDataPoints ? null : (_jsxs(_Fragment, { children: [customDataPoint ? (isWebApp ? (_jsx(ForeignObject, { height: svgHeight, width: totalWidth, x: initialSpacing -
                                    dataPointsWidth / 2 +
                                    ((_e = spacingArray[index - 1]) !== null && _e !== void 0 ? _e : 0), y: getYOrSecondaryY(item.y) - dataPointsHeight / 2, children: customDataPoint(item, index) })) : (_jsx(Animated.View, { style: {
                                    position: 'absolute',
                                    // height: svgHeight,
                                    // width: totalWidth,
                                    left: initialSpacing -
                                        dataPointsWidth / 2 +
                                        ((_f = spacingArray[index - 1]) !== null && _f !== void 0 ? _f : 0),
                                    top: getYOrSecondaryY(item.y) - dataPointsHeight / 2,
                                    opacity: isAnimated ? appearingOpacity : 1,
                                }, children: customDataPoint(item, index) }))) : null, dataPointsShape === 'rectangular' ? (_jsx(Fragment, { children: customDataPoint ? null : (_jsx(AnimatedRect, { x: getX(spacingArray, index) - dataPointsWidth / 2, y: getYOrSecondaryY(item.y) - dataPointsHeight / 2, width: isAnimated ? growingWidth[index] : dataPointsWidth, height: isAnimated ? growingHeight[index] : dataPointsHeight, opacity: isAnimated ? appearingDataPoints : opacity, fill: showDataPointOnFocus
                                        ? index === selectedIndex
                                            ? dataPointsColor
                                            : 'none'
                                        : dataPointsColor, onPress: function () {
                                        item.onPress
                                            ? item.onPress(item, index)
                                            : props.onPress
                                                ? props.onPress(item, index)
                                                : focusEnabled
                                                    ? handleFocus(index)
                                                    : null;
                                    }, onPressOut: function () {
                                        if (!item.onPress && !props.onPress && focusEnabled) {
                                            handleUnFocus();
                                        }
                                    } })) }, index)) : (_jsx(Fragment, { children: customDataPoint ? null : (_jsx(AnimatedCircle, { cx: getX(spacingArray, index), cy: getYOrSecondaryY(item.y), r: isAnimated ? growingRadii[index] : dataPointsRadius, fill: showDataPointOnFocus
                                        ? index === selectedIndex
                                            ? dataPointsColor
                                            : 'none'
                                        : dataPointsColor, opacity: isAnimated ? appearingDataPoints : opacity, stroke: (_g = item.borderColor) !== null && _g !== void 0 ? _g : borderColor, strokeWidth: (_h = item.borderWidth) !== null && _h !== void 0 ? _h : borderWidth, onPress: function () {
                                        item.onPress
                                            ? item.onPress(item, index)
                                            : props.onPress
                                                ? props.onPress(item, index)
                                                : focusEnabled
                                                    ? handleFocus(index)
                                                    : null;
                                    }, onPressOut: function () {
                                        if (!item.onPress && !props.onPress && focusEnabled) {
                                            handleUnFocus();
                                        }
                                    } })) }, index)), dataPointLabelComponent ? (!showTextOnFocus || index === selectedIndex ? (isWebApp ? (_jsx(ForeignObject, { height: svgHeight, width: dataPointLabelWidth, x: initialSpacing +
                                    (item.dataPointLabelShiftX ||
                                        props.dataPointLabelShiftX ||
                                        0) -
                                    dataPointLabelWidth / 2 +
                                    spacing * index, y: containerHeight +
                                    (item.dataPointLabelShiftY ||
                                        props.dataPointLabelShiftY ||
                                        0) -
                                    (item.y * containerHeight) / maxValue, children: showDataPointLabelOnFocus
                                    ? index === selectedIndex &&
                                        (focusTogether || key == selectedLineNumber)
                                        ? dataPointLabelComponent(item, index) // not pushed in latest release
                                        : null
                                    : dataPointLabelComponent(item, index) })) : (_jsx(View, { style: {
                                    position: 'absolute',
                                    height: svgHeight,
                                    width: dataPointLabelWidth,
                                    left: initialSpacing +
                                        (item.dataPointLabelShiftX ||
                                            props.dataPointLabelShiftX ||
                                            0) -
                                        dataPointLabelWidth / 2 +
                                        spacing * index,
                                    top: containerHeight +
                                        (item.dataPointLabelShiftY ||
                                            props.dataPointLabelShiftY ||
                                            0) -
                                        (item.y * containerHeight) / maxValue,
                                }, children: showDataPointLabelOnFocus
                                    ? index === selectedIndex &&
                                        (focusTogether || key == selectedLineNumber)
                                        ? dataPointLabelComponent(item, index) // not pushed in latest release
                                        : null
                                    : dataPointLabelComponent(item, index) }))) : null) : text || item.dataPointText ? (!showTextOnFocus || index === selectedIndex ? (_jsx(CanvasText, { fill: item.textColor || textColor, fontSize: item.textFontSize || textFontSize, x: getX(spacingArray, index) -
                                    dataPointsWidth +
                                    (item.textShiftX || props.textShiftX || 0), y: getYOrSecondaryY(item.y) -
                                    dataPointsHeight / 2 +
                                    (item.textShiftY || props.textShiftY || 0), children: !showTextOnFocus && !showValuesAsDataPointsText
                                    ? item.dataPointText
                                    : text })) : null) : null, index === selectedIndex ? _jsx(Text, { children: '' }) : null] }))] }, index));
        });
    };
    var svgWrapperViewStyle = {
        position: 'absolute',
        top: 0, // 281 + xAxisLabelsVerticalShift + labelsExtraHeight - xAxisThickness,
        left: 0,
        zIndex: 1,
        transform: [{ scaleX: I18nManager.isRTL ? -1 : 1 }],
    };
    var renderChartContent = function () {
        var _a;
        return (_jsxs(View, { style: [
                svgWrapperViewStyle,
                {
                    width: totalWidth,
                    height: containerHeightIncludingBelowXAxis,
                    //   zIndex,
                },
            ], children: [_jsx(Svg, { height: svgHeight, width: totalWidth, onPress: props.onBackgroundPress, children: renderDataPoints(hideDataPoints, props.data, props.data, dataPointsShape, dataPointsWidth, dataPointsHeight, dataPointsColor, dataPointsRadius, textColor, textFontSize, startIndex, endIndex, false, showValuesAsDataPointsText, cumulativeSpacing, 0) }), (_a = props.data) === null || _a === void 0 ? void 0 : _a.map(function (item, index) {
                    var _a, _b, _c, _d, _e, _f;
                    var secondaryLabel = (_c = (_a = item.secondaryLabel) !== null && _a !== void 0 ? _a : (_b = secondaryXAxis === null || secondaryXAxis === void 0 ? void 0 : secondaryXAxis.labelTexts) === null || _b === void 0 ? void 0 : _b[index]) !== null && _c !== void 0 ? _c : '';
                    var secondaryLabelTextStyle = (_f = (_e = (_d = item.secondaryLabelTextStyle) !== null && _d !== void 0 ? _d : secondaryXAxis === null || secondaryXAxis === void 0 ? void 0 : secondaryXAxis.labelsTextStyle) !== null && _e !== void 0 ? _e : item.labelTextStyle) !== null && _f !== void 0 ? _f : props.xAxisLabelTextStyle;
                    return (_jsxs(View, { children: [isAnimated
                                ? renderAnimatedLabel(false, index, item.label ||
                                    (props.xAxisLabelTexts && props.xAxisLabelTexts[index]
                                        ? props.xAxisLabelTexts[index]
                                        : ''), item.labelTextStyle || props.xAxisLabelTextStyle, item.labelComponent)
                                : renderLabel(false, index, item.label ||
                                    (props.xAxisLabelTexts && props.xAxisLabelTexts[index]
                                        ? props.xAxisLabelTexts[index]
                                        : ''), item.labelTextStyle || props.xAxisLabelTextStyle, item.labelComponent), secondaryXAxis
                                ? isAnimated
                                    ? renderAnimatedLabel(true, index, secondaryLabel, secondaryLabelTextStyle, item.secondaryLabelComponent)
                                    : renderLabel(true, index, secondaryLabel, secondaryLabelTextStyle, item.secondaryLabelComponent)
                                : null] }, index));
                })] }));
    };
    var remainingScrollViewProps = {
        onScroll: function (ev) {
            var _a;
            (_a = props.onScroll) === null || _a === void 0 ? void 0 : _a.call(props, ev);
        },
        bounces: props.bounces,
        overScrollMode: (_d = props.overScrollMode) !== null && _d !== void 0 ? _d : 'auto',
    };
    return (_jsx(BarAndLineChartsWrapper, __assign({}, barAndLineChartsWrapperProps, { scrollRef: scrollRef, animatedWidth: widthValue, renderChartContent: renderChartContent, remainingScrollViewProps: remainingScrollViewProps, nestedScrollEnabled: props.nestedScrollEnabled })));
};
