// gifted-charts-core/dist/BubbleChart/index.js
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { useState } from 'react';
import { computeMaxAndMinItems, getAxesAndRulesProps, getExtendedContainerHeightWithPadding, getMaxValue, getNoOfSections, indexOfFirstNonZeroDigit } from '../utils';
import { AxesAndRulesDefaults, BubbleDefaults, chartTypes, LineDefaults } from '../utils/constants';
export var useBubbleChart = function (props) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36;
    var _37 = props.data, data = _37 === void 0 ? [] : _37;
    var noOfSections = getNoOfSections(props.noOfSections, props.maxValue, props.stepValue);
    var containerHeight = (_a = props.height) !== null && _a !== void 0 ? _a : (((_b = props.stepHeight) !== null && _b !== void 0 ? _b : 0) * noOfSections ||
        AxesAndRulesDefaults.containerHeight);
    var valuesRange = Math.max.apply(Math, __spreadArray([], __read(data.map(function (i) { return Math.max(i.value, 0); })), false)) - // find the largest +ve number
     Math.min.apply(// find the largest +ve number
    Math, __spreadArray([], __read(data.map(function (i) { return Math.max(i.value, 0); })), false)); // find the smallest +ve number
    var showFractionalValues = (_c = props.showFractionalValues) !== null && _c !== void 0 ? _c : (isFinite(valuesRange) && valuesRange <= 1);
    var roundToDigits = (_d = props.roundToDigits) !== null && _d !== void 0 ? _d : (showFractionalValues ? indexOfFirstNonZeroDigit(valuesRange) + 1 : 0);
    var _38 = computeMaxAndMinItems(data, false, // extrapolateMissingValues
    roundToDigits, showFractionalValues, data), maxItem = _38.maxItem, minItem = _38.minItem;
    var maxValue = getMaxValue(props.maxValue, props.stepValue, noOfSections, maxItem) || 10;
    var mostNegativeValue = (_e = props.mostNegativeValue) !== null && _e !== void 0 ? _e : minItem;
    var onlyPositive = (_f = props.onlyPositive) !== null && _f !== void 0 ? _f : mostNegativeValue >= 0;
    var horizSections = [{ value: '0' }];
    var stepHeight = (_g = props.stepHeight) !== null && _g !== void 0 ? _g : containerHeight / noOfSections;
    var stepValue = (_h = props.stepValue) !== null && _h !== void 0 ? _h : maxValue / noOfSections;
    var noOfSectionsBelowXAxis = (_j = props.noOfSectionsBelowXAxis) !== null && _j !== void 0 ? _j : Math.round(Math.ceil(-mostNegativeValue / stepValue));
    var labelsExtraHeight = (_k = props.labelsExtraHeight) !== null && _k !== void 0 ? _k : 0;
    var yAxisLabelWidth = (_l = props.yAxisLabelWidth) !== null && _l !== void 0 ? _l : (props.hideYAxisText
        ? AxesAndRulesDefaults.yAxisEmptyLabelWidth
        : AxesAndRulesDefaults.yAxisLabelWidth);
    var horizontal = false;
    var yAxisAtTop = false;
    var initialSpacing = (_m = props.initialSpacing) !== null && _m !== void 0 ? _m : LineDefaults.initialSpacing;
    var xAxisThickness = (_o = props.xAxisThickness) !== null && _o !== void 0 ? _o : AxesAndRulesDefaults.xAxisThickness;
    var spacing = (_p = props.spacing) !== null && _p !== void 0 ? _p : LineDefaults.spacing;
    var cumulativeSpacing = [];
    var spacingSum = 0;
    var space = (_q = props.spacing) !== null && _q !== void 0 ? _q : spacing;
    data.forEach(function (item) {
        var _a;
        spacingSum += (_a = item.spacing) !== null && _a !== void 0 ? _a : space;
        cumulativeSpacing.push(spacingSum);
    });
    var totalWidth = initialSpacing +
        spacing * (data.length - 1) +
        ((_r = props.endSpacing) !== null && _r !== void 0 ? _r : LineDefaults.endSpacing);
    var endSpacing = (_s = props.endSpacing) !== null && _s !== void 0 ? _s : LineDefaults.endSpacing;
    var disableScroll = (_t = props.disableScroll) !== null && _t !== void 0 ? _t : false;
    var showScrollIndicator = (_u = props.showScrollIndicator) !== null && _u !== void 0 ? _u : false;
    var scrollToEnd = (_v = props.scrollToEnd) !== null && _v !== void 0 ? _v : false;
    var scrollAnimation = (_w = props.scrollAnimation) !== null && _w !== void 0 ? _w : false;
    var scrollEventThrottle = (_x = props.scrollEventThrottle) !== null && _x !== void 0 ? _x : 16;
    var focusEnabled = (_y = props.focusEnabled) !== null && _y !== void 0 ? _y : LineDefaults.focusEnabled;
    var showDataPointOnFocus = (_z = props.showDataPointOnFocus) !== null && _z !== void 0 ? _z : LineDefaults.showDataPointOnFocus;
    var _39 = __read(useState((_0 = props.focusedDataPointIndex) !== null && _0 !== void 0 ? _0 : -1), 2), selectedIndex = _39[0], setSelectedIndex = _39[1];
    var _40 = __read(useState(''), 2), points = _40[0], setPoints = _40[1];
    var _41 = __read(useState(-1), 2), pointerIndex = _41[0], setPointerIndex = _41[1];
    var _42 = __read(useState(0), 2), pointerX = _42[0], setPointerX = _42[1];
    var _43 = __read(useState(0), 2), pointerY = _43[0], setPointerY = _43[1];
    var isAnimated = (_1 = props.isAnimated) !== null && _1 !== void 0 ? _1 : false;
    var horizontalRulesStyle = props.horizontalRulesStyle;
    var yAxisOffset = (_2 = props.yAxisOffset) !== null && _2 !== void 0 ? _2 : 0;
    var showXAxisIndices = (_3 = props.showXAxisIndices) !== null && _3 !== void 0 ? _3 : AxesAndRulesDefaults.showXAxisIndices;
    var xAxisIndicesHeight = (_4 = props.xAxisIndicesHeight) !== null && _4 !== void 0 ? _4 : AxesAndRulesDefaults.xAxisIndicesHeight;
    var xAxisIndicesWidth = (_5 = props.xAxisIndicesWidth) !== null && _5 !== void 0 ? _5 : AxesAndRulesDefaults.xAxisIndicesWidth;
    var xAxisIndicesColor = (_6 = props.xAxisIndicesColor) !== null && _6 !== void 0 ? _6 : AxesAndRulesDefaults.xAxisIndicesColor;
    //   const pointerConfig = props.pointerConfig
    var getPointerProps = (_7 = props.getPointerProps) !== null && _7 !== void 0 ? _7 : null;
    var dataPointsRadius = (_8 = props.dataPointsRadius) !== null && _8 !== void 0 ? _8 : LineDefaults.dataPointsRadius;
    var dataPointsWidth = (_9 = props.dataPointsWidth) !== null && _9 !== void 0 ? _9 : LineDefaults.dataPointsWidth;
    var extraWidthDueToDataPoint = props.hideDataPoints
        ? 0
        : dataPointsRadius !== null && dataPointsRadius !== void 0 ? dataPointsRadius : dataPointsWidth;
    var xAxisLabelsAtBottom = (_10 = props.xAxisLabelsAtBottom) !== null && _10 !== void 0 ? _10 : false;
    var allowFontScaling = (_11 = props.allowFontScaling) !== null && _11 !== void 0 ? _11 : AxesAndRulesDefaults.allowFontScaling;
    var animationDuration = (_12 = props.animationDuration) !== null && _12 !== void 0 ? _12 : LineDefaults.animationDuration;
    var negativeStepValue = (_13 = props.negativeStepValue) !== null && _13 !== void 0 ? _13 : stepValue;
    var secondaryMaxValue = maxValue;
    var secondaryMinItem = maxValue * -1;
    var showSecondaryFractionalValues = false;
    var secondaryRoundToDigits = 1;
    var axesAndRulesProps = getAxesAndRulesProps(props, containerHeight, stepHeight, stepValue, noOfSections, roundToDigits, negativeStepValue, secondaryMaxValue, secondaryMinItem, showSecondaryFractionalValues, secondaryRoundToDigits);
    var fourthQuadrantHeight = noOfSectionsBelowXAxis * stepHeight;
    var extendedContainerHeight = getExtendedContainerHeightWithPadding(containerHeight, 0);
    var containerHeightIncludingBelowXAxis = extendedContainerHeight + fourthQuadrantHeight;
    var mostNegativeValueOnYAxis = negativeStepValue * noOfSectionsBelowXAxis;
    var containsNegativeValue = ((_14 = props.mostNegativeValue) !== null && _14 !== void 0 ? _14 : 0) < 0 ||
        ((_15 = props.data) === null || _15 === void 0 ? void 0 : _15.some(function (item) { var _a; return ((_a = item.y) !== null && _a !== void 0 ? _a : 0) < 0; }));
    var getX = function (spacingArray, index) {
        var _a, _b;
        return (_b = (_a = props.data) === null || _a === void 0 ? void 0 : _a[index].x) !== null && _b !== void 0 ? _b : initialSpacing + (index ? spacingArray[index - 1] : 0);
    };
    var getY = function (value) {
        if (containsNegativeValue && value < 0 && stepValue !== negativeStepValue) {
            return (extendedContainerHeight +
                (value * fourthQuadrantHeight) / mostNegativeValueOnYAxis);
        }
        return extendedContainerHeight - (value * containerHeight) / maxValue;
    };
    var showTextOnFocus = (_16 = props.showTextOnFocus) !== null && _16 !== void 0 ? _16 : LineDefaults.showTextOnFocus;
    var showDataPointLabelOnFocus = (_17 = props.showDataPointLabelOnFocus) !== null && _17 !== void 0 ? _17 : LineDefaults.showDataPointLabelOnFocus;
    var focusTogether = (_18 = props.focusTogether) !== null && _18 !== void 0 ? _18 : true;
    var _44 = __read(useState(-1), 2), selectedLineNumber = _44[0], setSelectedLineNumber = _44[1];
    var lastLineNumber = 1;
    // if (props.secondaryData) {
    //   lastLineNumber = 6667 // lastLineNumber is 6667 for a secondary line, so the index or key of the secondary line is 6666
    // }
    // if (props.data2) lastLineNumber = 2
    // if (props.data3) lastLineNumber = 3
    // if (props.data4) lastLineNumber = 4
    // if (props.data5) lastLineNumber = 5
    // if ((props.dataSet?.length ?? 0) > lastLineNumber)
    //   lastLineNumber = props.dataSet?.length ?? 0
    var unFocusOnPressOut = (_19 = props.unFocusOnPressOut) !== null && _19 !== void 0 ? _19 : LineDefaults.unFocusOnPressOut;
    var delayBeforeUnFocus = (_20 = props.delayBeforeUnFocus) !== null && _20 !== void 0 ? _20 : LineDefaults.delayBeforeUnFocus;
    var handleFocus = function (index) {
        setSelectedIndex(index);
    };
    var handleUnFocus = function () {
        if (unFocusOnPressOut) {
            setTimeout(function () { return setSelectedIndex(-1); }, delayBeforeUnFocus);
        }
    };
    var hideDataPoints = (_21 = props.hideDataPoints) !== null && _21 !== void 0 ? _21 : false;
    var dataPointsShape = (_22 = props.dataPointsShape) !== null && _22 !== void 0 ? _22 : LineDefaults.dataPointsShape;
    var dataPointsHeight = (_23 = props.dataPointsHeight) !== null && _23 !== void 0 ? _23 : LineDefaults.dataPointsHeight;
    var dataPointsColor = (_24 = props.dataPointsColor) !== null && _24 !== void 0 ? _24 : BubbleDefaults.dataPointsColor;
    var textFontSize = (_25 = props.textFontSize) !== null && _25 !== void 0 ? _25 : LineDefaults.textFontSize;
    var textColor = (_26 = props.textColor) !== null && _26 !== void 0 ? _26 : LineDefaults.textColor;
    var startIndex = 0;
    var endIndex = (_28 = (_27 = props.data) === null || _27 === void 0 ? void 0 : _27.length) !== null && _28 !== void 0 ? _28 : 0;
    var showValuesAsDataPointsText = (_29 = props.showValuesAsDataPointsText) !== null && _29 !== void 0 ? _29 : LineDefaults.showValuesAsDataPointsText;
    var xAxisLabelsVerticalShift = (_30 = props.xAxisLabelsVerticalShift) !== null && _30 !== void 0 ? _30 : (xAxisLabelsAtBottom
        ? fourthQuadrantHeight
        : AxesAndRulesDefaults.xAxisLabelsVerticalShift);
    var xAxisTextNumberOfLines = (_31 = props.xAxisTextNumberOfLines) !== null && _31 !== void 0 ? _31 : AxesAndRulesDefaults.xAxisTextNumberOfLines;
    var rotateLabel = (_32 = props.rotateLabel) !== null && _32 !== void 0 ? _32 : false;
    var borderWidth = (_33 = props.borderWidth) !== null && _33 !== void 0 ? _33 : BubbleDefaults.borderWidth;
    var borderColor = (_34 = props.borderColor) !== null && _34 !== void 0 ? _34 : BubbleDefaults.borderColor;
    var opacity = (_35 = props.opacity) !== null && _35 !== void 0 ? _35 : BubbleDefaults.opacity;
    var barAndLineChartsWrapperProps = {
        chartType: chartTypes.BUBBLE,
        containerHeight: containerHeight,
        noOfSectionsBelowXAxis: noOfSectionsBelowXAxis,
        stepHeight: stepHeight,
        negativeStepHeight: stepHeight,
        labelsExtraHeight: labelsExtraHeight,
        yAxisLabelWidth: yAxisLabelWidth,
        horizontal: horizontal,
        rtl: false,
        shiftX: 0,
        shiftY: 0,
        yAxisAtTop: yAxisAtTop,
        initialSpacing: initialSpacing,
        data: data,
        stackData: undefined, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        secondaryData: undefined, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        barWidth: 0, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        xAxisThickness: xAxisThickness,
        totalWidth: totalWidth,
        disableScroll: disableScroll,
        showScrollIndicator: showScrollIndicator,
        scrollToEnd: scrollToEnd,
        scrollToIndex: props.scrollToIndex,
        scrollAnimation: scrollAnimation,
        scrollEventThrottle: scrollEventThrottle,
        indicatorColor: props.indicatorColor,
        selectedIndex: [selectedIndex],
        setSelectedIndex: setSelectedIndex,
        spacing: spacing,
        showLine: false,
        lineConfig: null, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        lineConfig2: null, // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        maxValue: maxValue,
        lineData: [], // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        lineData2: [], // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        lineBehindBars: false,
        points: points,
        points2: '', // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        arrowPoints: [], // Not needed but passing this prop to maintain consistency (between LineChart and BarChart props)
        // horizSectionProps-
        width: props.width,
        horizSections: horizSections,
        endSpacing: endSpacing,
        horizontalRulesStyle: horizontalRulesStyle,
        noOfSections: noOfSections,
        sectionColors: props.sectionColors,
        showFractionalValues: showFractionalValues,
        axesAndRulesProps: axesAndRulesProps,
        yAxisLabelTexts: props.yAxisLabelTexts,
        yAxisOffset: yAxisOffset,
        rotateYAxisTexts: 0,
        hideAxesAndRules: props.hideAxesAndRules,
        showXAxisIndices: showXAxisIndices,
        xAxisIndicesHeight: xAxisIndicesHeight,
        xAxisIndicesWidth: xAxisIndicesWidth,
        xAxisIndicesColor: xAxisIndicesColor,
        getPointerProps: getPointerProps,
        pointerIndex: pointerIndex,
        pointerX: pointerX,
        pointerY: pointerY,
        onEndReached: props.onEndReached,
        onStartReached: props.onStartReached,
        endReachedOffset: (_36 = props.endReachedOffset) !== null && _36 !== void 0 ? _36 : LineDefaults.endReachedOffset,
        onMomentumScrollEnd: props.onMomentumScrollEnd,
        extraWidthDueToDataPoint: extraWidthDueToDataPoint,
        customBackground: props.customBackground,
        onlyPositive: onlyPositive,
        highlightEnabled: LineDefaults.highlightEnabled,
        lowlightOpacity: LineDefaults.lowlightOpacity,
        xAxisLabelsAtBottom: xAxisLabelsAtBottom,
        onScrollEndDrag: props.onScrollEndDrag,
        allowFontScaling: allowFontScaling,
        showVerticalLines: props.showVerticalLines
    };
    return {
        totalWidth: totalWidth,
        animationDuration: animationDuration,
        containerHeightIncludingBelowXAxis: containerHeightIncludingBelowXAxis,
        getY: getY,
        barAndLineChartsWrapperProps: barAndLineChartsWrapperProps,
        getX: getX,
        maxValue: maxValue,
        selectedIndex: selectedIndex,
        setSelectedIndex: setSelectedIndex,
        showTextOnFocus: showTextOnFocus,
        focusEnabled: focusEnabled,
        focusTogether: focusTogether,
        selectedLineNumber: selectedLineNumber,
        lastLineNumber: lastLineNumber,
        initialSpacing: initialSpacing,
        spacing: spacing,
        containerHeight: containerHeight,
        handleFocus: handleFocus,
        handleUnFocus: handleUnFocus,
        isAnimated: isAnimated,
        showDataPointOnFocus: showDataPointOnFocus,
        showDataPointLabelOnFocus: showDataPointLabelOnFocus,
        dataPointsShape: dataPointsShape,
        dataPointsWidth: dataPointsWidth,
        dataPointsHeight: dataPointsHeight,
        dataPointsColor: dataPointsColor,
        dataPointsRadius: dataPointsRadius,
        textColor: textColor,
        textFontSize: textFontSize,
        startIndex: startIndex,
        endIndex: endIndex,
        showValuesAsDataPointsText: showValuesAsDataPointsText,
        cumulativeSpacing: cumulativeSpacing,
        hideDataPoints: hideDataPoints,
        xAxisLabelsVerticalShift: xAxisLabelsVerticalShift,
        labelsExtraHeight: labelsExtraHeight,
        xAxisThickness: xAxisThickness,
        xAxisTextNumberOfLines: xAxisTextNumberOfLines,
        rotateLabel: rotateLabel,
        allowFontScaling: allowFontScaling,
        borderWidth: borderWidth,
        borderColor: borderColor,
        opacity: opacity
    };
};
